lib/graphql/schema/warden.rb

```diff
@@ -126,8 +126,44 @@ def root_type_for_operation(op_name)
 
       private
 
+      def union_memberships(obj_type)
+        @unions ||= read_through { |obj_type| @schema.union_memberships(obj_type).select { |u| visible?(u) } }
+        @unions[obj_type]
+      end
+
       def visible_field?(field_defn)
-        visible?(field_defn) && visible?(field_defn.type.unwrap)
+        visible?(field_defn) && visible_type?(field_defn.type.unwrap)
+      end
+
+      def visible_type?(type_defn)
```

Comment:
Quite a conditional, but basically a type should be visible if:

  - Its own visibility is `true`

and one of these:

  - It's a root type
  - It's an introspection type
  - its referenced by a field or argument
  - it has a visible abstract type (part of Union or Interface)
  - It is an abstract type and has visible members

Reply:
:+1: 😎 

It's a lot of syntax but for my part, I appreciate it!